internal/blocksync/reactor.go

```diff
@@ -557,7 +557,12 @@ func (r *Reactor) poolRoutine(ctx context.Context, stateSynced bool, blockSyncCh
 				// validate the block before we persist it
 				err = r.blockExec.ValidateBlock(ctx, state, first)
 			}
-
+			if err == nil && state.ConsensusParams.ABCI.VoteExtensionsEnabled(extCommit.Height) {
```

Comment:
I think there is an extra thing we need to do in the blocksync reactor. We need to modify the "request" and "reply" logic for blocks so that:
* if `state.ConsensusParams.ABCI.VoteExtensionsEnabled(extCommit.Height)` is true, we keep requiring the extended commit with extensions, both when building the reply and when receiving it
* else, we send `nil` as extension, and we drop the message upon reception if the extension is not `nil` (malformed message)

Reply:
Another thing we need to adapt is the condition in line 468 above . Now, things are more complex:
* if we are above "required extension height" then the condition is similar: at least one block needs to be blocksynched (to make sure we have the required extension commit) unless we are at `initialHeight` (which doesn't even need a commit)
* if we are below "required extension height", we can probably proceed without waiting for syncing one block

Reply:
> I think there is an extra thing we need to do in the blocksync reactor. We need to modify the "request" and "reply"

I'm not sure we need to know this data when we receive the block. We go own to verify the correctness of the block message we received within the `poolRoutine`, [here](https://github.com/tendermint/tendermint/blob/351b3bf6b0e775fe2a461d212ab2c6f7d166a102/internal/blocksync/reactor.go#L569). This would effectively be verifying parts of this response multiple times.


I've added logic to query the state store when the peer is making a request to us because we need to know if we should grab from the `commit` data or the `extendedcommit` data.

Reply:
> Another thing we need to adapt is the condition in line 468 above

adapted, although, is there a reason to use this `state.LastBlockHeight > 0` instead of the `stateSynced` param that the `poolRoutine` method receives?

Reply:
> block message we received within the poolRoutine, [here](https://github.com/tendermint/tendermint/blob/351b3bf6b0e775fe2a461d212ab2c6f7d166a102/internal/blocksync/reactor.go#L569)

I had missed that yesterday when I went through the PR. I'll check those changes now

Reply:
`stateSynced` would work as well. When I was considering which one to use, I leaned towards "self-contained" checks (so we don't depend of the correctness of data coming from elsewhere), but I realize it's not very strong as a reason.
I think both options are correct... if you want to switch the condition to `stateSynced`-based, I'm OK with that.

Reply:
I see what this was doing, got it. So, it is possible to blocksync a crash recovered node. I've added an updated check along with documentation that I think will make this a bit cleaner without having to check the db on each iteration of this loop, let me know if it makes sense to you.

Reply:
Updated the check, please take a look.